taichi/transforms/loop_invariant_code_motion.cpp

```diff
@@ -0,0 +1,158 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/statements.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/visitors.h"
+#include "taichi/system/profiler.h"
+
+#include <stack>
+
+TLANG_NAMESPACE_BEGIN
+
+class LoopInvariantCodeMotion : public BasicStmtVisitor {
+ public:
+  std::stack<Block *> loop_blocks;
+
+  const CompileConfig &config;
+
+  DelayedIRModifier modifier;
+
+  LoopInvariantCodeMotion(const CompileConfig &config) : config(config) {
+    allow_undefined_visitor = true;
+  }
+
+  bool stmt_can_be_moved(Stmt *stmt) {
+    if (loop_blocks.size() <= 1 || (!config.move_loop_invariant_outside_if &&
+                                    stmt->parent != loop_blocks.top()))
+      return false;
+
+    bool can_be_moved = true;
+
+    Block *current_scope = stmt->parent;
+
+    for (Stmt *operand : stmt->get_operands()) {
+      if (operand->parent == current_scope) {
+        // This statement has an operand that is in the current scope,
+        // so it can not be moved out of the scope.
+        can_be_moved = false;
+        break;
+      }
+      if (config.move_loop_invariant_outside_if &&
+          stmt->parent != loop_blocks.top()) {
+        // If we enable moving code from a nested if block, we need to check
+        // visibility. Example:
+        // for i in range(10):
+        //   a = x[0]
+        //   if b:
+        //     c = a + 1
+        // Since we are moving statements outside the cloest for scope,
+        // We need to check the scope of the operand
+        Stmt *operand_parent = operand;
+        while (operand_parent && operand_parent->parent) {
+          operand_parent = operand_parent->parent->parent_stmt;
+          if (!operand_parent) break;
+          // If the one of the parent of the operand is the top loop scope
+          // Then it will not be visible if we move it outside the top loop
+          // scope
+          if (operand_parent == loop_blocks.top()->parent_stmt) {
+            can_be_moved = false;
+            break;
+          }
+        }
+        if (!can_be_moved)
+          break;
+      }
+    }
+
+    return can_be_moved;
+  }
+
+  void visit(BinaryOpStmt *stmt) override {
+    if (stmt_can_be_moved(stmt)) {
+      auto replacement = stmt->clone();
+      stmt->replace_with(replacement.get());
+
+      modifier.insert_before(stmt->parent->parent_stmt, std::move(replacement));
```

Comment:
Shall `stmt->parent->parent_stmt` here be `loop_blocks.top()->parent_stmt`? If a statement is inside an `if` inside a `for`, it is moved out of the `if` but it is still inside the `for`. This may also be the reason why the performance is low when you enabled `move_loop_invariant_outside_if`.

Reply:
I'll fix it in #6072 and test again.

Reply:
Ah right makes sense, thanks for noticing it!

Reply:
Good catch! However, I think the result will be the same. For example, in this code,
```python
a = x[0]
for i in range(10):
  if b:
    c = a + 1
```
we will move `c = a + 1` like this:
```python
a = x[0]
for i in range(10):
  c = a + 1
  if b:
    pass
```
In the next iteration of this pass, we will have:
```python
a = x[0]
c = a + 1
for i in range(10):
  if b:
    pass
```
So this will only increase the number of iterations by 1 and only if this is the bottleneck (the statement that will be moved the most times).

Nevertheless, thanks for catching this!

Reply:
We might need a cost function to properly address this